fix: avoid killing process group when not applicable

Alizter opened the pull request:
We add a way of tracking whether a job is a process group leader to the
scheduler. This allows us to avoid sending a signal to a process group
when it isn't needed.

Doing so fixes an issue with dune exec -w where rebuilds would send a
kill signal to a non-existant process group assumed to be attached to
the process we were execing.

- [x] depends on #12361 
- [x] changelog
- fixes #12323

rgrinberg commented:
I don't think we should have any "fallbacks". If we are launching a process without a process group, we know we are doing that and there's no need to send a signal to the group

Alizter commented:
Sounds good to me. I will need to design a more principled solution that tracks the process group as you said. It might take some time before I can get round to it however.

Alizter committed:
fix: avoid killing process group when not applicable

We add a way of tracking whether a job is a process group leader to the
scheduler. This allows us to avoid sending a signal to a process group
when it isn't needed.

Doing so fixes an issue with dune exec -w where rebuilds would send a
kill signal to a non-existant process group assumed to be attached to
the process we were execing.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter commented:
:boom: I've messed up `killall` on Windows, the CI will not finished.

Alizter commented:
> Is this relevant to this issue?
> 
> ```
> Run opam exec -- make release
> ocamlc -output-complete-exe -intf-suffix .dummy -g -o .duneboot.exe -I boot -I +unix unix.cma boot/types.ml boot/libs.ml boot/duneboot.ml
> .\.duneboot.exe
> cd _boot && D:\a\dune\dune\_opam\bin\ocamlopt.opt.exe -c -g -I +unix -I +threads -ccopt -D_UNICODE -ccopt -DUNICODE mdb.c
> vendor/ocaml-lmdb\mdb.c: In function 'mdb_strerror':
> vendor/ocaml-lmdb\mdb.c:1809:1: warning: function may return address of local variable [-Wreturn-local-addr]
> vendor/ocaml-lmdb\mdb.c:1770:7: note: declared here
> ```

Those are just warnings when compiled on Windows, the actual issue is the fact that the runner took 5 hours.

rgrinberg commented:
Feel free to re-open if you intend to work on this again.

rgrinberg commented on src/dune_scheduler/scheduler.ml:
```diff
@@ -112,9 +112,9 @@ let with_job_slot f =
 ;;
 
-let wait_for_process t pid =
+let wait_for_process t ?(is_process_group_leader = true) pid =
```
Can you make this argument required?
